Preliminary work for transactional logic for our releases

AndreasBurger opened the pull request:
Restructures steps/release.py in such a way that each action with side effect is isolated from other actions with side effects by encapsulating them in classes.
Currently, this has the same behaviour as before but allows implementing revert/rollback logic for each side effect.

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,298 @@
+        return self._release_commit_sha
+
+
+class ReleaseProcessor(object):
```
how about: TransactionalStep? also add a _short_ coment

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,298 @@
+    def __init__(self, release_context: ReleaseContext):
+        self.release_context = release_context
+
+    def validate(self):
```
add @abc.abstractmethod

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+    Instances of this class are intended to perform exactly one change in their apply-method
+    and provide the corresponding undo-functionality in the revert-method
+    '''
+    def __init__(self, release_context: ReleaseContext):
```
so. as I elaborated above. the release-ctx should be named (and crafted) more generically. either by changing as described before. or by (this would be the easies approach) changing it to a generic object, such as a dict or a SimpleNamespace.

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+
+class ReleaseCommitStep(TransactionalStep):
+    def __init__(
+        self,
```
you could also add `*args, **kwargs` and pass this to the super-c'tor

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+    def name(self):
+        pass
+
+
```
I have already mentioned this in the documentation proposal. I suggest to add a class representing a sequence of steps. Maybe call it `Transaction`. It should receive the step instances and offer a method to run all steps in a transactional fashion (which you inlined a but further down).

That way (with the changes proposed up to here), we could consider to mv this out of this module and re-use it for other steps. This is however a change for another PR (the moving-out and re-use. the Transaction class should be introduced here).

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+    def validate(self):
+        pass
+
+    def apply(self):
```
I suppose apply is fine. Would it be more adequate to call it 'run' or 'execute' instead? Not sure, though. This is just a gut feeling.

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        return "Release Commit Creator"
+
+    def validate(self):
+        info(f'validating { self.name()}')
```
mv any printing to the transaction class (for all steps)

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        self.release_commit_callback = release_commit_callback
+
+    def name(self):
+        return "Release Commit Creator"
```
might it make more sense to choose names that describe the action (e.g. 'create release commit')

Then, we would have error messages like:
"Operation 'create release commit' failed"

(or status messages like "Operation 'create release commit' succeeded")

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        )
+
+    def revert(self):
+        # TODO: revert commit
```
how about omitting the unimplemented method and print a warning instead (in the base class)?
at any rate: do not return silently

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        self.author_email = not_empty(author_email)
+
+    def validate(self):
+        info(f'validating { self.name()}')
```
mv to transation

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+            )
+
+    def name(self):
+        return "Release Tag Creator"
```
create release tag

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        )
+
+    def revert(self):
+        # TODO: Delete tag
```
ditto - do not silently return

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        git_helper: GitHelper,
+        githubrepobranch: GitHubRepoBranch,
+        release_version: str,
+        release_context: ReleaseContext,
```
the ctx should be either at the beginning or at the end of the argument list, as it is somewhat "inherited". but maybe, we will not need to pass it like this anyway

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        self.component_descriptor_file_path = existing_file(component_descriptor_file_path)
+
+    def name(self):
+        return "Release Creator"
```
create github release

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+    def apply(
+        self,
+    ):
+        info('Create release')
```
mv to tx

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+            )
+
+    def revert(self):
+        #TODO: remove release from GitHub
```
do not return silently

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,19 +28,303 @@
+        self.prerelease_suffix = not_empty(prerelease_suffix)
+
+    def name(self):
+        return "Next Cycle Commit Creator"
```
prepare next dev cycle

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +29,317 @@
+        info(f'validating {self.name()}')
+         # ensure the given callback script exists
+        if self.release_commit_callback:
+            self.release_commit_callback = os.path.join(
```
no! this is a very very bad idea. either omit this in the validation, or mv the code to a method that returns the next version. then call it in `validate` w/o storing it. and call it again in `apply` (persisting it this time).

In case I missed other ocurrences: never ever modify the step object (or anything else) during validation. This is super-bad! Any step *must* successfully execute (at least potentially) w/o running validate before

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +29,317 @@
+            version_str=self.release_version,
+            operation=self.version_operation,
+        )
+        self.next_cycle_dev_version = version.process_version(
```
actually, I meant to place the "this is superbad" comment here. but the diff was updated before i could save. so as mentioned earlier: NEVER modify state during validation

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +29,317 @@
+        )
+
+    def revert(self):
+        #TODO: remove created commit
```
do not return silently

ccwienk commented on concourse/steps/release.py:
```diff
@@ -49,166 +350,98 @@ def release_and_prepare_next_dev_cycle(
+        p.validate()
+
+    successful_steps = list()
+    for step in steps:
```
firstly, mv this into the transaction class (see above).
secondly, you also must handle any exceptions occurring during reversion. we must try to revert as much as possible, even if errors occur in the process

ccwienk commented on concourse/steps/release.py:
```diff
@@ -49,166 +350,98 @@ def release_and_prepare_next_dev_cycle(
+            step.revert()
+            # undo all previous changes in reverse order
+            successful_steps.reverse()
+            for s in successful_steps:
```
nit: use reversed(...)

AndreasBurger commented:
I've implemented the current suggestions/requested changes:
- I've created the suggested `Transaction` class and moved the 'old' release context there after generalising it a bit more (as suggested) to a transaction context. An instance of this object will be provided to each transaction.
- I've removed all state-modifying validation. Now, simple sanity checks are performed in the constructor (e.g. no empty string passed) whereas more sophisticated checks (e.g.: Is the given string *actually* a semver) are performed in the validate() method. The release-note verification, however, had to be moved to the apply() method of the relevant step. Might be worth it to move it to the constructor, actually. I'm not sure yet.

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +30,392 @@
+        # validate type of args and set context
+        for step in steps:
+            if not isinstance(step, TransactionalStep):
+                raise RuntimeError('Transactions may only contain instances of TransactionalStep')
```
nit: ValueError would be more appropriate

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +30,392 @@
+            step_name = step.name()
+            # attempt to execute the step
+            try:
+                info(f"Executing step '{step_name}'")
```
execute or apply? ;-)

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +30,392 @@
+                info(f"Reverting step {step_name}")
+                s.revert()
+            except BaseException as e:
+                all_reverted = False
```
this is in general fine. it might help us to also dump the stacktrace in these cases (`traceback.print_exc()`). Admittedly, this is not very sophisticated. but it tends to be helpful for error analysis..

AndreasBurger commented:
Done. In detail:
- Unimplemented revert-methods now raise an UnimplementedMethodError
- Release notes are now posted in their own three-step-transaction (Update release notes for the GitHub release -> cleanup old draft release -> post slack release notes [optionally]). Steps in this transaction won't actually do anything when reverting, resulting in a 'succesful' release job with a warning if release notes couldn't be posted, but the release was created. 
Also of note here is commit 98e44daf2616fa5fd08b5157901c9233864db340 since it introduces an 'update release notes' method to the GitHubHelper. It should be `cherry-pick`able to the main branch if it is in order.
- I've cleaned up the commit history some more after the previous review. All changes to the Transaction's revert logic are now included in its initial commit. No sense in producing refactoring commits in a PR.

ccwienk commented on test/concourse/steps/release_test.py:
```diff
@@ -0,0 +1,257 @@
+        def constructor(
+            slack_cfg_name='test_config',
+            slack_channel='test_channel',
+            githubrepobranch=MagicMock(),
```
please create real objects here. there is no need to use a mock, since this is just a plain python object contain some strings, is it not?

ccwienk commented on test/concourse/steps/release_test.py:
```diff
@@ -0,0 +1,257 @@
+
+    def test_validation_fails_on_invalid_semver(self, examinee):
+        with pytest.raises(ValueError):
+            examinee(release_version='made_up_version').validate()
```
could you use names such as 'invalid_version' to indicate that the value is intentionally invalid? same for file system paths. the name (if possible to be arbitrarily chosen) should reflect the intention.

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +31,508 @@
+            step_name = step.name()
+            # attempt to execute the step
+            info(f"Applying step '{step_name}'")
+            executed_steps.append(step)
```
add after successful execution?

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +31,508 @@
+        )
+        if release_commit_callback:
+            self.release_commit_callback = str(
+                pathlib.Path(repo_dir_absolute, release_commit_callback)
```
why create a pythlib object just to convert it back into a str? just use os.path.join.

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +31,508 @@
+        self.release_version = not_empty(release_version)
+        self.rebase_before_release = rebase_before_release
+
+        repo_dir_absolute = pathlib.Path(not_empty(repo_dir)).resolve()
```
as already pointed out. please just use `str`. OR replace all usages at least in this module with pathlib.Paths. it is certainly not a good idea to mix styles. and since we need to interface with mako, str will be the more approprate type..

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +31,494 @@
+            author_name=self.author_name,
+            author_email=self.author_email
+        )
+        return {}
```
?

ccwienk commented on concourse/steps/release.py:
```diff
@@ -23,22 +31,508 @@
+                    label=product.model.COMPONENT_DESCRIPTOR_ASSET_NAME,
+                )
+
+        return {}
```
why return an empty dict?
